src/service/invoice-pdf-service.ts

```diff
@@ -0,0 +1,236 @@
+      }),
+      products:products,
+      pricing: pricing,
+      subject: invoice.description,
+      // Are unused but still required.
+      sender: new Identity({ firstName: UNUSED_PARAM, fullName: UNUSED_PARAM, lastName: UNUSED_PARAM, lastNamePreposition: UNUSED_PARAM }),
+      recipient: new Identity({ firstName: UNUSED_PARAM, lastName: UNUSED_PARAM, lastNamePreposition: UNUSED_PARAM, fullName: UNUSED_PARAM }),
+
+      dates: new Dates({
+        date: invoice.createdAt,
+      }),
+      company: new Company({
+        name: invoice.addressee,
+      }),
+      address: new Address({
+        street: invoice.street,
+        postalCode: invoice.postalCode,
+        city: invoice.city,
+        country: invoice.country,
+      }),
+    });
+  }
+
+  /**
+   * Prepares and returns the parameters required by the PDF generator client to create an invoice PDF.
+   * This includes setting up file settings and consolidating invoice parameters.
+   *
+   * @param {Invoice} invoice - The invoice for which to generate PDF parameters.
+   * @returns {InvoiceRouteParams} An instance of `InvoiceRouteParams` containing the consolidated parameters and settings for PDF generation.
+   */
+  static getPdfParams(invoice: Invoice): InvoiceRouteParams {
+    const params = this.getInvoiceParameters(invoice);
+
+    const settings: FileSettings = new FileSettings({
+      createdAt: new Date(),
+      fileType: ReturnFileType.PDF,
+      language: Language.ENGLISH,
+      name: '',
+      stationery: 'BAC',
+    });
+
+    const data: IInvoiceRouteParams = {
+      params,
+      settings,
+    };
+
+    return new InvoiceRouteParams(data);
+  }
+
+  /**
+   * Generates a PDF for an invoice and uploads it to the file service. If the invoice PDF generation or upload fails, it throws an error with the failure reason.
+   *
+   * @param {number} invoiceId - The ID of the invoice to generate and upload the PDF for.
+   * @param {PdfGenerator} pdfGenerator - The services used for generating the PDF and handling file operations.
+   * @returns {Promise<InvoicePdf>} A promise that resolves to the `InvoicePdf` entity representing the generated and uploaded PDF.
+   */
+  public static async createInvoicePDF(invoiceId: number, pdfGenerator: PdfGenerator): Promise<InvoicePdf> {
+    const invoice = await Invoice.findOne({ where: { id: invoiceId }, relations: ['to', 'invoiceStatus', 'transfer', 'transfer.to', 'transfer.from', 'pdf', 'invoiceEntries'] });
+    if (!invoice) return undefined;
+
+    const params = this.getPdfParams(invoice);
+    return pdfGenerator.client.generateInvoice(InvoiceType.Invoice, params).then(async (res: FileResponse) => {
+      const blob = res.data;
+      const buffer = Buffer.from(await blob.arrayBuffer());
+      return pdfGenerator.fileService.uploadInvoicePdf(invoice, buffer, invoice.to, hashJSON(this.getInvoiceParameters(invoice)));
+    }).catch((res: any) => {
+      throw new Error(`Invoice generation failed: ${res}`);
```

Comment:
According to the coverage report, this catch and throw has not been covered. It is however quite important as you mentioned at the bar a moment ago, because the user should get feedback in case the generator is failing to do its job.

Reply:
should be fixed